packages/astro/src/@types/astro.ts

```diff
@@ -1660,12 +1660,47 @@ export type PaginateFunction = (data: any[], args?: PaginateOptions) => GetStati
 
 export type Params = Record<string, string | undefined>;
 
+export type SupportsKind = 'Unsupported' | 'Stable' | 'Experimental' | 'Deprecated';
```

Comment:
Can you explain the reason for this?

Reply:
I think `undefined` is inevitably going to become a `SupportsKind`. For example, when a new capability is added and a user updates astro but not the adapter.

Reply:
@matthewp, I thought we could have some flexibility in the future with the features that expose to the adapters. Here's what I had in mind:

- Unsupported -> needed to throw an error
- Experimental -> a way to tell the user that the feature is new and might be buggy/change. The user will see a terminal log
- Deprecated -> to gracefully tell the user that some feature might be removed from the adapter
- Stable -> to tell Astro that a feature is supported

Reply:
Yeah, like @lilnasy said we are going to have to support `undefined` here. IIRC the reason to have this feature is so that when we add new features we can tell if an adapter is supporting it.

To that end, I don't think any adapter is going to set `Unsupported` on purpose. More likely it just never gets set at all until they come to support it.

So I'm wondering if having these different options might be premature. Would it not be better to make it a boolean `false | true` and default it to `false`? Then we could add additional flags like `Deprecated` in the future. But I'm not sure if we should add them until we have a use-case for them.

Reply:
> Yeah, like @lilnasy said we are going to have to support undefined here. IIRC the reason to have this feature is so that when we add new features we can tell if an adapter is supporting it.

`Unsupported` will be exactly `undefined`. Use case: we release the `build.split` feature, and it's only added to the node adapter. Then, a user tries to use the feature for - for example - the deno adapter. Astro will throw an error because no value has been set, and the `Unsupported` is the default value. (I will have to fix the assignment of the default value). 

> To that end, I don't think any adapter is going to set Unsupported on purpose. More likely it just never gets set at all until they come to support it.

Actually, there are cases where features are **intetionally unsupported**. For example Vercel Edge won't support the edge middleware, so it makes sense to explicitly flag `Unsupported`. Another example is Cloudflare, which doesn't support `"static"` output.